<!DOCTYPE html>
<html lang="en">
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

    <!-- Metro 4 -->
    <link rel="stylesheet" href="https://cdn.metroui.org.ua/v4/css/metro-all.min.css">
</head>
<body class="mx-auto p-3">
    <h1>Jasbir Bhangoo</h1>

    <h3>Genome Assembly and Alignment</h3>

    <h6>Summary</h6>
    <p>DNA sequencers "read" short segments or "sequences" from the genome. The challenge is to re-assemble them into the full genome.
        Sequencers are imperfect. They contain errors and gaps. Also the collection of all sequences are not complete.
        Therefore the same genome must be sequenced many times to introduce redundancy. But this makes it hard to re-assemble the genome because short
        segments are often repeated in multiple places in the genome. It is hard to tell if any particular repeated sequence is
        a copy of one previously seen or from a different place.
    </p>
    <p>
        These examples were generated from exercises for Coursera classes
        <a href="https://www.coursera.org/learn/dna-sequencing">Algorithms for DNA Sequencing</a> and
        <a href="https://www.coursera.org/learn/genome-sequencing?specialization=bioinformatic">Genome Sequencing (Bioinformatics II)</a>
    </p>

    <h6>Read Assembly</h6>
    <p>
        Two methods are implemented: Overlapping Reads and DeBruijn Network.
    </p>
    <div class="row">
        <div class="col-xl-6 col-lg-6 col-md-6 col-sm-6 col-xs-6">
            <span style="font-weight:500;">Overlapping reads summary</span>
            <ol>
            <li>Extract all k-mers from all sequencing reads. Keep a data structure of all reads that each k-mer was seen in
            <li>Use the data structure to create a candidate list of all read pairs with a k-mer in common
            <li>Test each candidate read pair and build an adjascency graph from all read pairs that overlap.
            (A read pair overlaps if some suffix of the first is a prefix of the second)
            <li>Determine the contigs
            <ol>
            <li>Trace all paths through the adjasency graph
            <li>For each path, assemble the contig by merging the suffix of each node with the overlapping prefix of the next
            </ol>
            </ol>
        </div>
        <div class="col-xl-6 col-lg-6 col-md-6 col-sm-6 col-xs-6">
            <span style="font-weight:500;">De Bruijn summary</span>
            <ol>
            <li>Extract all k-mers from all sequencing reads
            <li>Split each k-mer into left and right (k-1)-mers
            <li>Create an edge from the left part to the right part. The resulting directed mutli-graph is a
                <span style="font-style:oblique;">De Bruijn Network</span>.
            <li>Try to find an <span style="font-style:oblique;">Euler Path</span> through the network
                that passes through each edge exactly once. This would be the reconstruction of the original sequence.
                <li>In practice, the reads are imperfect. So allow for errors by combining (k-1)-mers that are similar
                into one node. Return the longest path through the network if the Euler Path does not exist.
            </ol>
        </div>
    </div>
    <h6>Paired Read Assembly</h6>
    <p>
    <span style="font-style:oblique;">Paired-end</span> DNA sequencers produce longer much sequences, but they have a
        "gap" region in the middle with unknown composition. Therefore, each read consists of two sequences whose
        position is known relative to each other.
    </p>
    <h6>Source Code</h6>
    <p>
        Program code can be check out from the associated
        <a href="https://github.com/jbhangoo/jbhangoo.github.io/tree/master/src/Genome">Github repository</a>.
    </p>
    <h6>Data Source</h6>
    <p>A repository of genome data is publicly available from the
    <a href="https://www.ncbi.nlm.nih.gov/guide/howto/dwn-genome/">National Center for Biotechnology Information</a>.
    Unless ownership rights are specifically claimed, they can be used freely.
    </p>
</body>
</html>